### PR TITLE
fix: DSF increase on checkpointed nonzero shares

### PR DESCRIPTION
**Pull request by ypatil12**
**Motivation:**

There exists an edge case where:

1. Alice queues a withdrawal for all her shares
2. Alice's operator is slashed
3. Alice checkpoints (no rewards or slashes). This actually increases her DSF, even though there wasn't a balance increase on the beacon chain
4. Alice completes withdrawal as tokens
5. Alice redeposits

After step 5, Alice's withdrawable shares are overcounted due to increasing the DSF.  Note that in step 4 if Alice completes her withdrawal as shares she can still inflate her withdrawable shares. 

**Modifications:**

The fix is to prevent any state updates if `addedShares` is 0. In addition, we also update `recordBeaconChainETHBalanceUpdate` to only call the DM if there is a nonzero `balanceDeltaWei`, which is symmetric to the behavior in the `StrategyManager`. 

Regression tests:

- [x] testFuzz_noCall_zeroBalanceUpdate
- [x] testFuzz_deposit_delegate_allocate_slash_checkpointZeroBalance
- [x] testFuzz_deposit_delegate_allocate_slashAndQueue_checkpoint_redeposit
- [x] testFuzz_deposit_delegate_allocate_slashAndQueue_completeAsTokens_redeposit
- [x] testFuzz_deposit_delegate_allocate_slashAndQueue_checkPoint_completeAsShares

**Result:**

Inflating withdrawable shares edge case fixed.

### CODE REVIEW

**Review by 8sunyuan — DISMISSED** [review]

File: src/test/integration/tests/Slashed_Eigenpod_AVS.t.sol [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+        // 10. Redeposit
+        cheats.deal(address(staker), 32 ether);
+        (uint40[] memory newValidators, uint64 addedBeaconBalanceGwei) = staker.startValidators();
+        beaconChain.advanceEpoch_NoRewards();
+        staker.verifyWithdrawalCredentials(newValidators);
+        check_VerifyWC_State(staker, newValidators, addedBeaconBalanceGwei);
+    }
+
+    /// @dev Asserts that the staker cannot inflate withdrawable shares after checkpointing & completing as shares
```
Comment (0xClandestine) [inline src/test/integration/tests/Slashed_Eigenpod_AVS.t.sol:145]: _randBool with asTokens path would be a nice to have

**Review by 0xClandestine — CHANGES_REQUESTED** [review]

File: src/test/integration/tests/Slashed_Eigenpod_AVS.t.sol (hunk 1 above)
Comment (ypatil12) [reply src/test/integration/tests/Slashed_Eigenpod_AVS.t.sol:145]: The issue here is that we do a separate flow if it's as tokens or as shares

**Review by 0xClandestine — APPROVED** [review]

**Review by wadealexc — DISMISSED** [review]

**Review by nadir-akhtar — DISMISSED** [review]

**Review by 8sunyuan — APPROVED** [review]
